astropy/table/table.py

```diff
@@ -335,6 +362,88 @@ def __get__(self, instance, owner_cls):
         return info
 
 
+class HiddenColumns(MetaAttribute):
+    """Maintain tuple of table column names that are hidden from print output.
+
+    This is a descriptor that inherits from MetaAttribute so that the attribute
+    value is stored in the table meta['__attributes__'].
+    """
+    def __get__(self, instance, owner_cls):
+        # For getting from class not an instance
+        if instance is None:
+            return self
+
+        hidden_columns = instance.__dict__.get('hidden_columns')
+        if hidden_columns is None:
+            hidden_columns = deepcopy(self)
+            instance.__dict__['hidden_columns'] = hidden_columns
+        # We set _parent_ref on every call, since if one makes copies of
+        # instances, 'hidden_columns' will be copied as well, which will lose the
+        # reference.
+        hidden_columns._instance_ref = weakref.ref(instance)
+        return hidden_columns
+
+    def __set__(self, instance, names):
+        if isinstance(names, str):
+            names = [names]
+        if names is None:
+            # Remove hidden_columns from the meta['__attributes__'] dict.
+            # Subsequent access will just return None.
+            del instance.hidden_columns
+        else:
+            return super().__set__(instance, tuple(names))
+
+    def __call__(self):
+        instance = self._instance_ref()
+        return super().__get__(instance, instance.__class__)
+
+    def __repr__(self):
+        if hasattr(self, '_instance_ref'):
+            out = f'<{self.__class__.__name__} name={self.name} value={self()}>'
+        else:
+            out = super().__repr__()
+        return out
+
+    def _add_remove_setup(self, names):
+        if isinstance(names, str):
+            names = [names]
+        instance = self._instance_ref()
+        value = super().__get__(instance, instance.__class__)
+        if value is None:
+            value = ()
+        return instance, names, value
+
+    def add(self, names):
+        instance, names, value = self._add_remove_setup(names)
+        value = tuple(set(value) | set(names))
+        super().__set__(instance, value)
+
+    def remove(self, names):
+        instance, names, value = self._add_remove_setup(names)
+        set_value = set(value)
+        set_names = set(names)
+        if set_names - set_value:
+            missing = ', '.join(set_names - set_value)
+            raise ValueError(f'{missing} not in hidden_columns')
```

Comment:
Following `set()`, `KeyError` may be more appropriate.

Reply:
I actually started with that then discovered this interesting behavior with `KeyError` that put me off. It displays the `repr` of the argument, not the `str`. I think somehow the intent is that you should raise `KeyError(key)`, which doesn't quite fit here. But I can go either way.
```
In [10]: raise ValueError('my message')                                                                                             
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-a8f0b130d67a> in <module>
----> 1 raise ValueError('my message')

ValueError: my message

In [11]: raise KeyError('my message')                                                                                               
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-11-0aef9f386bb7> in <module>
----> 1 raise KeyError('my message')

KeyError: 'my message'
```